xtm/xtm_api.h

```diff
@@ -0,0 +1,214 @@
+ * @param[in] queue - xtm_queue to delete.
+ * @retval    0 on success. Otherwise -1 with errno set appropriately.
+ *            (as in close(2), since it implies close of fds).
+ */
+int
+xtm_queue_delete(struct xtm_queue *queue);
+
+/**
+ * Notify queue consumer.
+ * @param[in] queue - xtm_queue to notify.
+ * @retval    0 on success. Otherwise -1 with errno set appropriately.
+ *            (as in write(2), since it implies write to internal fd).
+ */
+int
+xtm_queue_notify_consumer(struct xtm_queue *queue);
+
+/**
+ * Notify queue producer, when queue is not full.
+ * @param[in] queue - xtm_queue to notify.
+ * @retval    0 on success. Otherwise -1 with errno set appropriately.
+ *            (as in write(2), since it implies write to internal fd).
+ */
+int
+xtm_queue_notify_producer(struct xtm_queue *queue);
+
+/**
+ * Check is there are free space in queue.
+ * @param[in] queue - xtm_queue to ckeck free space.
+ * @retval    0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+ */
+int
+xtm_queue_probe(struct xtm_queue *queue);
+
+/**
+ * Return current count of data in xtm queue.
+ * @param[in] queue - xtm_queue to check current count of data.
+ * @retval    count of data in queue.
+ */
+unsigned
+xtm_queue_count(struct xtm_queue *queue);
+
+/**
+ * Push function and it's argument to the queue. This function does not
+ * notify the consumer thread, but only push to the queue. To notify the
+ * consumer thread you must call xtm_queue_notify_consumer. The less often
+ * you notify, the greater the performance, but the greater the latency.
+ * @param[in] queue   - xtm_queue to push.
+ * @param[in] fun     - function to push.
+ * @param[in] fun_arg - function argument to push.
+ * @retval    0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+ */
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
+		   void *fun_arg);
+
+/**
+ * Return file descriptor, that should be watched by consumer thread to
+ * become readable. When it became readable, consumer should call one
+ * of the consumer functions: xtm_queue_pop_ptrs or xtm_queue_invoke_funs_all.
+ * @param[in] queue - xtm_queue to get file descriptor.
+ * @retval    xtm queue file descriptor for consumer thread.
+ */
+int
+xtm_queue_consumer_fd(struct xtm_queue *queue);
+
+/**
+ * Return file descriptor, that should be watched by producer thread to
+ * become readable. When it became readable, producer may push to queue,
+ * being sure that there is a place in it (there is a little race between
+ * consumer and producer thread, so there is a little chance, that this
+ * descriptor become readable, but there is no free space in the queue. In
+ * this case producer thread must poll this descriptor again.
+ * @param[in] queue - xtm_queue to get file descriptor.
+ * @retval    xtm queue file descriptor for producer thread.
+ */
+int
+xtm_queue_producer_fd(struct xtm_queue *queue);
+
+/**
+ * Calls all functions contained in the queue.
```

Comment:
You definitely should add a comment about the necessity to notify the producer.
xtm_queue_pop_ptrs() also.
readme also.

Reply:
Fixed